dapr/proto/runtime/v1/appcallback.proto

```diff
@@ -74,6 +74,18 @@ message TopicEventRequest {
 
 // TopicEventResponse is response from app on published message
 message TopicEventResponse {
+  // TopicEventResponseStatus allows apps to have finer control over handling of the message.
```

Comment:
we do need to have the same change for http api, don't we?

Reply:
I would say no, because we use the http native status codes which is natural for the protocol. With gRPC we are taking the natural way for proto which is explicit model types.

Reply:
Grpc can return status code with the detail error info just like http. If we do not have a plan to have the same response, then using error status of grpc makes sense.

Reply:
It can, but it's not as natural for the developer with gRPC to return an error code as it is for HTTP 1.1.

Reply:
Can you elaborate more on the grpc and http api behaviors? 

Let's say that if message is dropped.
* For HTTP, dapr returns 500(or the other error codes) with the event deliver status
* For GRPC, dapr returns Status.OK with `TopicEventResponse`

http api returns error while grpc api returns ok. I do not think it make sense. 

Honestly, I do not understand the meaning of `natural`.

Reply:
Protocol-specific status codes (i.e. HTTP Status codes) are... well, specific to the protocol. gRPC uses uses a set of well-defined [status codes](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md) that are part of the RPC API for status. If we want to include a more detail, that can be used to drive Dapr-specific behaviour, I propose we build these into the returned object.

Reply:
@youngbupark do we have an actual example where this happens? Internal server error in HTTP and OK (0) in gRPC?

Reply:
> If we want to include a more detail, that can be used to drive Dapr-specific behaviour, I propose we build these into the returned object.

I agree on this. the point of my comment is that we should not have the different response status (success or fail). generally, both http and grpc apis need to returns the same success or fail response code.

"http returns fails while grpc returns OK" doesn't seem right.

Reply:
plus, success or fail response will make distributed trace telemetry differently per http or grpc.

Reply:
HTTP now has several status codes, 500, 503, 200 and 204.

The question is "what's the easiest, most natural experience for developers for the language and protocol they use"?

In HTTP 1.1, status codes are a common and accepted way to tell the caller about how you want it to respond to you.
In gRPC, status error codes are mostly used by the RPC framework itself to provide feedback to the caller about network failures/issues. It's less common for an application to use status codes to communicate intent for the caller, as opposed to its usage with HTTP. This is a major part of why gRPC status codes are also [limited](https://grpc.github.io/grpc/core/md_doc_statuscodes.html).

Reply:
Agree, so just to make sure we are on the same page, are you recommending we change the behaviour in HTTP so it returns status codes only in case of transport issues and include a response status struc in the payload which can be then used to consistently communicate operation status across gRPC and HTTP?

Reply:
Thanks for the explanation. (BTW, there are still error code to represent application behavior on client side: https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md. We used this for service invocation response code conversion. grpc-gateway project uses this way too)

Just to be clear,  referring to [http appcallback implementation](https://github.com/dapr/dapr/blob/b22d2125073fd4117a13c261dd7c179f8586733e/pkg/runtime/runtime.go#L1063-L1078), dapr's behavior for pubsub appcallback is:
* 200, 204 -> success
* 5xx -> retry
* < 500 and > 599 (3xx, 4xx) -> drop

(we need to document this behavior on api doc - created the issue https://github.com/dapr/docs/issues/772 to follow it up)

In gRPC of the current PR, dapr's behavior for pubsub appcallback is:
* OK with SUCCESS -> success
* non-OK status code, OK with RETRY -> retry
* OK with DROP -> drop.

If we wants to use this way, then we need to decide how to record metrics status label and trace result.

I.e. This results in the different metric and trace status code between http and gRPC. gRPC trace records success metrics and trace for retry and drop behavior while http records failure metrics and trace.

we need to either document this different behavior or correct it in dapr metrics and trace middleware.

Reply:
What do you recommend? Will changing it in the middleware constitute a breaking change?

Reply:
Here is my proposal: https://github.com/dapr/dapr/issues/1340#issuecomment-686714348

I believe we should also change the HTTP handling of retriable vs drop since not every 5xx is retriable and not every 4xx is a drop.

Reply:
> What do you recommend? Will changing it in the middleware constitute a breaking change?

it will not be breaking change but having exception for the specific api looks hacky. but, I am ok with either handling status code or OK statuscode + response body in http and gRPC API. it will resolve metrics and trace issues. having consistent response makes people less confused in the long term.

Reply:
BTW - we should probably treat *all* 2XX status codes as success. 201 and 202 are in common use and have the differences between those and 200 are things that Dapr can ignore. None of the other 2XX's are really in use, but they all mean success.

Reply:
> BTW - we should probably treat _all_ 2XX status codes as success. 201 and 202 are in common use and have the differences between those and 200 are things that Dapr can ignore. None of the other 2XX's are really in use, but they all mean success.

the middleware implementation for metrics and trace treats all 2xx status code as a success now.